anomalib/models/rkde/region_extractor.py

```diff
@@ -0,0 +1,139 @@
+
+Region Extractor.
+"""
+
+# Copyright (C) 2022 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+from enum import Enum
+from typing import List
+
+import torch
+from torch import Tensor, nn
+from torchvision.models.detection import fasterrcnn_resnet50_fpn
+from torchvision.ops import boxes as box_ops
+
+from anomalib.data.utils.boxes import scale_boxes
+
+
+class RoiStage(str, Enum):
+    """Processing stage from which rois are extracted."""
+
+    RCNN = "rcnn"
+    RPN = "rpn"
+
+
+class RegionExtractor(nn.Module):
+    """Extracts regions from the image.
+
+    Args:
+        stage (RoiStage, optional): Processing stage from which rois are extracted.
+        score_threshold (float, optional): Mimumum confidence score for the region proposals.
+        min_size (int, optional): Minimum size in pixels for the region proposals.
+        iou_threshold (float, optional): Intersection-Over-Union threshold used during NMS.
+        max_detections_per_image (int, optional): Maximum number of region proposals per image.
+    """
+
+    def __init__(
+        self,
+        stage: RoiStage = RoiStage.RCNN,
+        score_threshold: float = 0.001,
+        min_size: int = 25,
+        iou_threshold: float = 0.3,
+        max_detections_per_image: int = 100,
+    ) -> None:
+        super().__init__()
+
+        # Affects global behaviour of the region extractor
+        self.stage = stage
+        self.min_size = min_size
+        self.iou_threshold = iou_threshold
+        self.max_detections_per_image = max_detections_per_image
+
+        # Affects behaviour depending on roi stage
+        rpn_top_n = max_detections_per_image if self.stage == RoiStage.RPN else 1000
+        rpn_score_thresh = score_threshold if self.stage == RoiStage.RPN else 0.0
+
+        # Create the model
+        self.faster_rcnn = fasterrcnn_resnet50_fpn(
+            pretrained=True,
+            rpn_post_nms_top_n_test=rpn_top_n,
+            rpn_score_thresh=rpn_score_thresh,
+            box_score_thresh=score_threshold,
+            box_nms_thresh=1.0,  # this disables nms (we apply custom label-agnostic nms during post-processing)
+            box_detections_per_img=1000,  # this disables filtering top-k predictions (we apply our own after nms)
+        )
+
+    @torch.no_grad()
+    def forward(self, batch: Tensor) -> Tensor:
+        """Forward pass of the model.
+
+        Args:
+            input (Tensor): Batch of input images of shape [B, C, H, W].
+
+        Raises:
+            ValueError: When ``stage`` is not one of ``rcnn`` or ``rpn``.
+
+        Returns:
+            Tensor: Predicted regions, tensor of shape [N, 5] where N is the number of predicted regions in the batch,
+                 and where each row describes the index of the image in the batch and the 4 bounding box coordinates.
+        """
+        if self.training:
+            raise ValueError("Should not be in training mode")
+
+        if self.stage == "rcnn":
```

Comment:
If we add `__str__` to the enum class then we can do `self.stage == RoiStage.RCNN`.
I tried this and it works
![image](https://user-images.githubusercontent.com/17232914/210395881-fe9aaa28-f750-42bc-b751-87d342ab2b7e.png)